Fix initPool method in QueryCtx.h

isadikov opened the pull request:
This is a follow-up of https://github.com/facebookincubator/velox/pull/3754. Unblocks https://github.com/prestodb/presto/pull/18966.

The `initPool` method is not idempotent and would attach a memory usage tracker every time the function is invoked which is a change in behaviour compared to the previous code.

This PR fixes the issue by only attaching a memory usage tracker for uninitialised pool only. I manually imported the commit in Presto to verify the change.

isadikov commented:
There does not seem to be a file with tests for QueryCtx.h so I did not add a test - this is implicitly tested in Presto.

isadikov commented:
@aditi-pandit Could you take a look? Thank you.

@marxhxxx Could you also review this PR since I am not very familiar with purpose of the original patch?

xiaoxmeng commented on velox/core/QueryCtx.h:
```diff
@@ -142,8 +142,8 @@ class QueryCtx : public Context {
     if (!pool_) {
```
```
if (pool_ == nullptr) {
   return;
}
...
```

xiaoxmeng replied:
Thanks for finding out. I am going to fix this plus the other issue in https://github.com/facebookincubator/velox/pull/3815 which is required by Meta internal use case.

xiaoxmeng reviewed (approved):
@isadikov thanks for the fix!

Yuhta commented:
Curious since this is private and only called in ctor, when is the ctor being called more than once on the same object?

xiaoxmeng commented:
> Curious since this is private and only called in ctor, when is the ctor being called more than once on the same object?

PrestoCpp has already set memory pool with a memory tracker limit. A previous change will cause this using unlimited memory tracker as we override the checker in the pool.

isadikov commented:
@xiaoxmeng I have addressed your comment. Would you be able to review again? Thanks.
Alternatively, we could keep the https://github.com/facebookincubator/velox/commit/79e3d80378ea1883c9873097e94cb43debb8905a commit in this PR, so let me know what version you would like to keep.

isadikov commented:
This has been fixed as part of https://github.com/facebookincubator/velox/pull/3815.
